tests/FSharp.Compiler.ComponentTests/Language/CopyAndUpdateTests.fs

```diff
@@ -0,0 +1,371 @@
+﻿module FSharp.Compiler.ComponentTests.Language.CopyAndUpdateTests
+
+open Xunit
+open FSharp.Test.Compiler
+open StructuredResultsAsserts
+
+[<Fact>]
+let ``Cannot update the same field twice in nested copy-and-update``() =
+    FSharp """
+type NestdRecTy = { B: string }
+
+type RecTy = { D: NestdRecTy; E: string option }
+
+let t2 x = { x with D.B = "a"; D.B = "b" }
+    """
+    |> withLangVersionPreview
+    |> typecheck
+    |> shouldFail
+    |> withDiagnostics [
+        (Error 668, Line 6, Col 21, Line 6, Col 22, "The field 'B' appears twice in this record expression or pattern")
+    ]
+
+[<Fact>]
+let ``Cannot use nested copy-and-update in lang version70``() =
+    FSharp """
+type NestdRecTy = { B: string }
+
+type RecTy = { D: NestdRecTy; E: string option }
+
+let t2 x = { x with D.B = "a" }
+    """
+    |> withLangVersion70
+    |> typecheck
+    |> shouldFail
+    |> withDiagnostics [
+        (Error 3350, Line 6, Col 21, Line 6, Col 24, "Feature 'Nested record field copy-and-update' is not available in F# 7.0. Please use language version 'PREVIEW' or greater.")
+    ]
+
+[<Fact>]
+let ``Nested copy-and-update merges same level updates``() =
+    FSharp """
+module CopyAndUpdateTests
+
+type AnotherNestedRecTy = { A: int }
+
+type NestdRecTy = { B: AnotherNestedRecTy; C: string }
+
+type RecTy = { D: NestdRecTy; E: string option }
+
+let t2 x = { x with D.B.A = 1; D.C = "ads" }
+    """
+    |> withLangVersionPreview
+    |> withNoDebug
+    |> withOptimize
+    |> compile
+    |> shouldSucceed
+    |> verifyIL [
+(*
+        public static CopyAndUpdateTests.RecTy t2(CopyAndUpdateTests.RecTy x)
+        {
+            return new CopyAndUpdateTests.RecTy(new CopyAndUpdateTests.NestdRecTy(new CopyAndUpdateTests.AnotherNestedRecTy(1), "ads"), x.E@);
+        }
+*)
+        """
+.method public static class CopyAndUpdateTests/RecTy 
+        t2(class CopyAndUpdateTests/RecTy x) cil managed
+{
+    
+  .maxstack  8
+  IL_0000:  ldc.i4.1
+  IL_0001:  newobj     instance void CopyAndUpdateTests/AnotherNestedRecTy::.ctor(int32)
+  IL_0006:  ldstr      "ads"
+  IL_000b:  newobj     instance void CopyAndUpdateTests/NestdRecTy::.ctor(class CopyAndUpdateTests/AnotherNestedRecTy,
+                                                                          string)
+  IL_0010:  ldarg.0
+  IL_0011:  ldfld      class [FSharp.Core]Microsoft.FSharp.Core.FSharpOption`1<string> CopyAndUpdateTests/RecTy::E@
+  IL_0016:  newobj     instance void CopyAndUpdateTests/RecTy::.ctor(class CopyAndUpdateTests/NestdRecTy,
+                                                                     class [FSharp.Core]Microsoft.FSharp.Core.FSharpOption`1<string>)
+  IL_001b:  ret
+} 
+        """
+    ]
+
+[<Fact>]
+let ``Nested copy-and-update correctly updates fields in nominal record``() =
```

Comment:
Thinking out loud - can we imagine any corner cases when the nested record is declared as mutable within the outer record?
I would assume this feature does not record and is agnostic to mutability (would do a new copy anyway, right?), but better to be safe.

Reply:
No, since field mutation and copy-and-update are mutually exclusive (can't do both as a single expression), we should be good here. If there's a mutable field somewhere in the nested path (long identifier), the enclosing record will be copied.